com.mlapi.contrib.extensions/Runtime/NetworkDiscovery/NetworkDiscovery.cs

```diff
@@ -1,58 +1,283 @@
+        using var reader = PooledNetworkReader.Get(buffer);
+        SetBufferContent(buffer, segment);
+
+        try
+        {
+            if (ReadAndCheckHeader(reader, MessageType.Response) == false)
+            {
+                return;
+            }
+
+            var receivedResponse = new TResponse();
+            receivedResponse.NetworkSerialize(reader.Serializer);
+            ResponseReceived(udpReceiveResult.RemoteEndPoint, receivedResponse);
+        }
+        catch (Exception e)
+        {
+            Debug.LogException(e);
+        }
+    }
+
+    async Task ReceiveBroadcastAsync()
+    {
+        UdpReceiveResult udpReceiveResult = await m_Client.ReceiveAsync();
+
+        var segment = new ArraySegment<byte>(udpReceiveResult.Buffer, 0, udpReceiveResult.Buffer.Length);
+
+        // Create a network buffer and put the received data into it
+        using var buffer = PooledNetworkBuffer.Get();
+        using var reader = PooledNetworkReader.Get(buffer);
+        SetBufferContent(buffer, segment);
+
+        try
+        {
+            if (ReadAndCheckHeader(reader, MessageType.BroadCast) == false)
+            {
+                return;
+            }
+
+            var receivedBroadcast = new TBroadCast();
+            receivedBroadcast.NetworkSerialize(reader.Serializer);
+
+            if (ProcessBroadcast(udpReceiveResult.RemoteEndPoint, receivedBroadcast, out TResponse response))
+            {
+                buffer.BitPosition = 0;
+                using (PooledNetworkWriter writer = PooledNetworkWriter.Get(buffer))
+                {
+                    WriteHeader(writer, MessageType.Response);
+
+                    response.NetworkSerialize(writer.Serializer);
+                    var data = new ArraySegment<byte>(buffer.GetBuffer(), 0, (int) buffer.Length);
+
+                    await m_Client.SendAsync(data.Array, data.Count, udpReceiveResult.RemoteEndPoint);
+                }
+            }
+        }
+        catch (Exception e)
+        {
+            Debug.LogException(e);
+        }
+    }
+
+    private static void SetBufferContent(NetworkBuffer buffer, ArraySegment<byte> arraySegment)
+    {
+        using (PooledNetworkWriter writer = PooledNetworkWriter.Get(buffer))
+        {
+            writer.WriteBytes(arraySegment.Array, arraySegment.Count);
+        }
+
+        buffer.BitPosition = 0;
+    }
+
+    private void WriteHeader(NetworkWriter writer, MessageType messageType)
+    {
+        // Serialize unique application id to make sure packet received is from same application.
+        writer.WriteInt64(m_UniqueApplicationId);
+
+        // Write a flag indicating whether this is a broadcast
+        writer.WriteByte((byte) messageType);
+    }
+
+    private bool ReadAndCheckHeader(NetworkReader reader, MessageType expectedType)
+    {
+        var receivedApplicationId = reader.ReadInt64();
+        if (receivedApplicationId != m_UniqueApplicationId)
+        {
+            return false;
```

Comment:
Do we want to add any Debug.Log here in case this fails? Same for the condition below?

Reply:
I don't think we should do that because we listen for all UDP messages which under some circumstances can be a lot. Some routers / devices do regular broadcasts.

Reply:
Cool